src/digests.c

```diff
@@ -0,0 +1,492 @@
+            return CKR_OK;
+        }
+    }
+    return CKR_MECHANISM_INVALID;
+}
+
+CK_RV p11prov_digest_get_by_name(const char *name, CK_MECHANISM_TYPE *digest)
+{
+    for (int i = 0; digest_map[i].digest != CK_UNAVAILABLE_INFORMATION; i++) {
+        for (int j = 0; digest_map[i].names[j] != NULL; j++) {
+            if (OPENSSL_strcasecmp(name, digest_map[i].names[j]) == 0) {
+                *digest = digest_map[i].digest;
+                return CKR_OK;
+            }
+        }
+    }
+    return CKR_MECHANISM_INVALID;
+}
+
+struct p11prov_digest_ctx {
+    P11PROV_CTX *provctx;
+    CK_MECHANISM_TYPE mechtype;
+
+    P11PROV_SESSION *session;
+};
+
+typedef struct p11prov_digest_ctx P11PROV_DIGEST_CTX;
+
+#define DISPATCH_DIGEST_NEWCTX_FN(mech, digest) \
+    static void *p11prov_##digest##_newctx(void *provctx) \
+    { \
+        P11PROV_DIGEST_CTX *dctx = OPENSSL_zalloc(sizeof(P11PROV_DIGEST_CTX)); \
+        if (dctx == NULL) { \
+            return NULL; \
+        } \
+        dctx->provctx = provctx; \
+        dctx->mechtype = mech; \
+        dctx->session = CK_INVALID_HANDLE; \
+        return dctx; \
+    }
+
+DISPATCH_DIGEST_NEWCTX_FN(CKM_SHA_1, sha1);
+DISPATCH_DIGEST_NEWCTX_FN(CKM_SHA224, sha224);
+DISPATCH_DIGEST_NEWCTX_FN(CKM_SHA256, sha256);
+DISPATCH_DIGEST_NEWCTX_FN(CKM_SHA384, sha384);
+DISPATCH_DIGEST_NEWCTX_FN(CKM_SHA512, sha512);
+DISPATCH_DIGEST_NEWCTX_FN(CKM_SHA512_224, sha512_224);
+DISPATCH_DIGEST_NEWCTX_FN(CKM_SHA512_256, sha512_256);
+DISPATCH_DIGEST_NEWCTX_FN(CKM_SHA3_224, sha3_224);
+DISPATCH_DIGEST_NEWCTX_FN(CKM_SHA3_256, sha3_256);
+DISPATCH_DIGEST_NEWCTX_FN(CKM_SHA3_384, sha3_384);
+DISPATCH_DIGEST_NEWCTX_FN(CKM_SHA3_512, sha3_512);
+DISPATCH_DIGEST_COMMON_FN(dupctx);
+DISPATCH_DIGEST_COMMON_FN(freectx);
+DISPATCH_DIGEST_COMMON_FN(init);
+DISPATCH_DIGEST_COMMON_FN(update);
+DISPATCH_DIGEST_COMMON_FN(final);
+DISPATCH_DIGEST_COMMON_FN(gettable_params);
+
+static void *p11prov_digest_dupctx(void *ctx)
+{
+    P11PROV_DIGEST_CTX *dctx = (P11PROV_DIGEST_CTX *)ctx;
+    P11PROV_DIGEST_CTX *newctx;
+    CK_SLOT_ID slotid = CK_UNAVAILABLE_INFORMATION;
+    CK_SESSION_HANDLE sess = CK_INVALID_HANDLE;
+    CK_FUNCTION_LIST *f;
+    CK_BYTE_PTR state = NULL;
+    CK_ULONG state_len;
+    CK_RV ret;
+
+    P11PROV_debug("digest dupctx, ctx=%p", ctx);
+
+    if (dctx == NULL) {
+        return NULL;
+    }
+
+    newctx = OPENSSL_zalloc(sizeof(P11PROV_DIGEST_CTX));
+    if (dctx == NULL) {
+        return NULL;
+    }
+    newctx->provctx = dctx->provctx;
+    newctx->mechtype = dctx->mechtype;
+
+    if (dctx->session == NULL) {
+        goto done;
+    }
+
+    /* This is not really funny. OpenSSL by default assumes contexts with
+     * operations in flight can be easily duplicated, with all the
+     * cryptographic status and then both context can keep going
+     * independently. We'll try here, but on failure we just 'move' the
+     * to the new token and hope for the best */
```

Comment:
We just move *what* to the new token? The session?

Reply:
Ah I should have written, we "move the session to the new context" ... thanks for the catch I will reword.